src/core/Rule.ml

```diff
@@ -259,6 +278,22 @@ and extract_reduction = Separate | Concat [@@deriving show]
 and extract_transform = NoTransform | Unquote | ConcatJsonArray
 [@@deriving show]
 
+(*****************************************************************************)
+(* Join mode *)
+(*****************************************************************************)
+
+type mode_for_join = Step_search of formula | Step_taint of taint_spec
+[@@deriving show]
+
+type step_info = {
+  step_formula : mode_for_join;
```

Comment:
you could also use [ search_mode | taint_mode ] 
and call it step_mode instead of step_formula